File: 0053_maximum_subarray/max_subarray.cpp
#include <algorithm> // *max_element()
#include <iostream>
#include <vector>

using namespace std;

/**
 * Time: O(n)
 * Space: O(1)
 */
int max_subarray(vector<int> &nums)
{
    if (!nums.size())
        return 0;
    int max_so_far = nums[0];
    int max_curr = nums[0];
    for (int i = 1; i < nums.size(); i++)
    {
        max_curr = max(nums[i], nums[i] + max_curr);
        max_so_far = max(max_so_far, max_curr);
    }
    return max_so_far;
}

/**
 * Time: O(n)
 * Space: O(1)
 */
int max_subarray_kadane(vector<int> &nums)
{
    if (!nums.size())
        return 0;
    for (int i = 1; i < nums.size(); i++)
        if (nums[i - 1] > 0)
            nums[i] = nums[i] + nums[i - 1];
    return *max_element(nums.begin(), nums.end());
}

File: 0053_maximum_subarray/max_subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
    return 0 if nums.empty?
    max_curr = max_so_far = nums[0]
    nums[1..].each do |n|
        max_curr = [n, max_curr + n].max
        max_so_far = [max_curr, max_so_far].max
    end
    max_so_far
end